### PR TITLE
userspace and system call handler code cleanups

### PR DESCRIPTION
**Pull request by tyrantlizard77**
Bunch of cleanups to userspace.c and the object validation test case.

The war on boilerplate for system call handlers marches on, it is now far simpler to specify handlers, and some common cases are now one-liners.

### CODE REVIEW

File: kernel/mutex.c [hunk 1]
```diff
@@ -82,12 +82,9 @@ void _impl_k_mutex_init(struct k_mutex *mutex)
 }
 
 #ifdef CONFIG_USERSPACE
-u32_t _handler_k_mutex_init(u32_t mutex, u32_t arg2, u32_t arg3,
-			    u32_t arg4, u32_t arg5, u32_t arg6, void *ssf)
+_SYSCALL_HANDLER1(k_mutex_init, mutex)
```
Comment (lpereira) [inline kernel/mutex.c]: k_mutex syscalls could be _SIMPLE_VOID.

File: kernel/sem.c [hunk 2]
```diff
@@ -191,32 +179,12 @@ int _impl_k_sem_take(struct k_sem *sem, s32_t timeout)
...
-			   u32_t arg4, u32_t arg5, u32_t arg6, void *ssf)
+_SYSCALL_HANDLER2(k_sem_take, sem, timeout)
 {
-	_SYSCALL_ARG1;
-
-	_SYSCALL_OBJ(sem_ptr, K_OBJ_SEM, ssf);
-	_impl_k_sem_reset((struct k_sem *)sem_ptr);
-
-	return 0;
+	_SYSCALL_OBJ(sem, K_OBJ_SEM);
```
Comment (lpereira) [inline kernel/sem.c]: This can also be a _SIMPLE syscall

File: kernel/stack.c [hunk 3]
```diff
@@ -56,15 +56,14 @@ void _impl_k_stack_init(struct k_stack *stack, u32_t *buffer, int num_entries)
...
 {
 	int num_entries = (int)num_entries_p;
 
 	/* FIXME why is 'num_entries' signed?? */
-	_SYSCALL_VERIFY(num_entries > 0, ssf);
-	_SYSCALL_OBJ_INIT(stack, K_OBJ_STACK, ssf);
-	_SYSCALL_MEMORY_ARRAY_WRITE(buffer, num_entries, sizeof(u32_t), ssf);
+	_SYSCALL_VERIFY(num_entries > 0);
+	_SYSCALL_OBJ_INIT(stack, K_OBJ_STACK);
+	_SYSCALL_MEMORY_ARRAY_WRITE(buffer, num_entries, sizeof(u32_t));
```
Comment (lpereira) [inline kernel/stack.c]: _SYSCALL_MEMORY_ARRAY_WRITE() will perform the multiplication and overflow check as if num_entries were unsigned.  Can this be a problem?

File: kernel/thread.c [hunk 4]
```diff
@@ -136,12 +132,8 @@ void *_impl_k_thread_custom_data_get(void)
 }
 
 #ifdef CONFIG_USERSPACE
-u32_t _handler_k_thread_custom_data_get(u32_t arg1, u32_t arg2, u32_t arg3,
-					u32_t arg4, u32_t arg5, u32_t arg6,
-					void *ssf)
+_SYSCALL_HANDLER0(k_thread_custom_data_get)
```
Comment (lpereira) [inline kernel/thread.c]: Could be a SIMPLE syscall

File: kernel/thread.c [hunk 5]
```diff
@@ -119,13 +119,9 @@ void _impl_k_thread_custom_data_set(void *value)
...
 #ifdef CONFIG_USERSPACE
-u32_t _handler_k_thread_custom_data_set(u32_t arg1, u32_t arg2, u32_t arg3,
-					u32_t arg4, u32_t arg5, u32_t arg6,
-					void *ssf)
+_SYSCALL_HANDLER1(k_thread_custom_data_set, data)
 {
-	_SYSCALL_ARG1;
-
-	_impl_k_thread_custom_data_set((void *)arg1);
+	_impl_k_thread_custom_data_set((void *)data);
```
Comment (lpereira) [inline kernel/thread.c]: Could be a SIMPLE syscall

File: kernel/timer.c [hunk 6]
```diff
@@ -280,3 +254,15 @@ s32_t _timeout_remaining_get(struct _timeout *timeout)
 	irq_unlock(key);
 	return __ticks_to_ms(remaining_ticks);
 }
+
+#ifdef CONFIG_USERSPACE
+_SYSCALL_HANDLER1_SIMPLE(k_timer_remaining_get, K_OBJ_TIMER, struct k_timer *);
+_SYSCALL_HANDLER1_SIMPLE(k_timer_user_data_get, K_OBJ_TIMER, struct k_timer *);
+
+_SYSCALL_HANDLER2(k_timer_user_data_set, timer, user_data)
```
Comment (lpereira) [inline kernel/timer.c]: Could be a SIMPLE_VOID

**Review by lpereira — COMMENTED** [review]
This indeed removes a lot of boilerplate. I think we shouldn't try to simplify this further by obscuring things with more macros -- this, IMO, is as obscure as it should get.

The only simplification that could be done -- and it's really nothing pressing -- is to try to have only one _SYSCALL_HANDLER macro that takes in a variable number of arguments, and expansion is performed by the preprocessor.  But this can get hairier than what's currently in place.

**Review by agross-oss — APPROVED** [review]
I was wondering about having it collapse down further as well, but this looks decent (w/ leandro's suggestions)

File: kernel/timer.c (hunk 6 above)
Comment (andrewboie) [reply kernel/timer.c]: no, it takes 2 parameters

**Comment by lpereira** [comment]
Took a stab at having only one _SYSCALL_HANDLER(), and have the preprocessor expand it to the appropriate macro: https://hastebin.com/ehajimitij.cpp

File: kernel/thread.c (hunk 5 above)
Comment (andrewboie) [reply kernel/thread.c]: no, data is not a kernel object but an arbitrary pointer

File: kernel/sem.c (hunk 2 above)
Comment (andrewboie) [reply kernel/sem.c]: no, because of the additional timeout parameter

File: kernel/stack.c (hunk 3 above)
Comment (andrewboie) [reply kernel/stack.c]: i don't know. I would prefer to look into this in another patch set.

File: kernel/thread.c (hunk 4 above)
Comment (andrewboie) [reply kernel/thread.c]: done

File: kernel/mutex.c (hunk 1 above)
Comment (andrewboie) [reply kernel/mutex.c]: k_mutex_lock takes an additional timeout parameter

**Comment by andrewboie** [comment]
I can't address all of these, some of these are not appropriate for the SIMPLE macros because they take additional arguments.
Unless you think I should introduce SIMPLE2, SIMPLE3, etc where the args are just passed along verbatim?

**Comment by lpereira** [comment]
No, I don't think that we should have more _SIMPLE macros.  Adding more of them will actually make things not simple.

**Comment by andrewboie** [comment]
Fixed k_thread_custom_data_get
Fixed valid checkpatch.pl error about test_generic_object()
Need an override for other checkpatch issues, they are false-positive

File: kernel/stack.c (hunk 3 above)
Comment (lpereira) [reply kernel/stack.c]: Alright.

**Review by lpereira — APPROVED** [review]

File: kernel/stack.c (hunk 3 above)
Comment (andrewboie) [reply kernel/stack.c]: I fixed this in #4302

**Review by nashif — APPROVED** [review]

**Pull request merged**